Added UTM and MGRS formats support

strump opened the pull request:
Working on issue #5204 .

Conversion formulas took from https://github.com/Turbo87/utm (lan,lon ↔ UTM) and [proj4js](https://github.com/proj4js/mgrs) (UTM ↔ MGRS).

## Android
Added two new coordinate formats: UTM and MGRS. Added label to understand which format is shown.

| Coordinates | Copy menu |
| ---- | ---- |
| ![place-page-screenshots](https://github.com/organicmaps/organicmaps/assets/720808/63392754-9c2d-46f1-8e3a-4d3830b4c176) | ![place-page-copy-menu](https://github.com/organicmaps/organicmaps/assets/720808/06f7cd87-279a-49ec-8cd5-1e5cdb6b9d0a) |

Some coordinates are not convertable to UTM or MGRS ( `lat <= -80` or `lat > 84`). In such cases `'N/A'` value is shown:
<img src="https://github.com/organicmaps/organicmaps/assets/720808/a38288b6-9df3-4f3a-a9a8-107bd4118931" width="300px"/>


## Search
UTM and MGRS formats are parsed to lat,lon in search field:

<img src="https://github.com/organicmaps/organicmaps/assets/720808/34ce204d-af66-4af9-a89e-d97138542a7e" alt="search" width="300px"/>

## Testing
Generated some test cases manually. And took sample dataset https://s3.amazonaws.com/mgrs.io/mgrsToGeo_WE.txt and converted to test cases. <s>Because of big number of cases files `utm_mgrs_utils_tests.cpp` and `utm_mgrs_coords_match_test.cpp` became huge. Should I keep all test cases?</s> Picked subset of those test cases.

## TODO
* iPhone place page doesn't show new coordinates. Need another PR to add all new coordinates to iOS.

strump commented on base/string_utils.hpp:
```diff
@@ -566,6 +566,9 @@ inline std::string to_string(uint64_t i) { return std::to_string(i); }
 //@}
 
+// Get string with fixed count of digits. Extra '0' are added at the begining to fit size.
+std::string to_string_prec(long l, int prec);
```
I implemented long to string conversion. Am I reinventing the wheel here?

strump commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,250 @@
+    return utm_mgrs_utils::MGRStoLatLon((double)easting, (double)northing, zone_code, zone_letter, square_code, lat, lon);
+}
+
+bool parseLong(string const & txt, long & value)
```
String to long conversion. Am I reinventing the wheel here?

vng replied:
strings::to_int64

biodranik commented:
If tests are fast, you can store them in the same txt file and load tests from it instead of hard-coding. Or leave them in the code. Or clean up similar tests and leave only unique cases.

vng commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+    }
+
+    return northing;
+}
```
Well, ok here, but please setup your editor:
- use 2 spaces for _tab_ instead of 4
- use C++ braces, from the new line, not K&R style
- do not use braces in _one line_ expressions

strump replied:
Reformatted code.

vng commented on search/processor.cpp:
```diff
@@ -619,6 +620,26 @@ bool Processor::SearchCoordinates()
 
+  {
+    double lat;
+    double lon;
```
I'd prefer to declare double lat, lon; once on top. And remove useless block braces {}

strump replied:
Removed extra `{}` and don't declare `double lat, lon` multiple times.

vng commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,251 @@
+#include <iostream>
+#include <math.h>
+
+using namespace std;
```
Put this using inside namespace search. Or don't use using at all.

strump replied:
Moved into namespace.

vng commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,251 @@
+{
+
+bool IsSpace(char ch);
+string::size_type FindNonSpace(string const & txt, string::size_type startPos);
```
Well, string::size_type is a useless overkill (but with correct semantics). Use size_t instead.

strump replied:
Replaced with `size_t`.

vng commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,251 @@
+        northing *= pow(10L, decShift);
+    }
+
+    return utm_mgrs_utils::MGRStoLatLon((double)easting, (double)northing, zone_code, zone_letter, square_code, lat, lon);
```
So MGRStoLatLon should take integer or float values by design?

strump replied:
When I parse UTM or MGRS I use only integers. E.g. `"15 N 500000 4649776"`. But after parsing is done conversion from UTM/MRGS parameters to lat, lot is done using double values.

vng replied:
I say that probably better to take long input easting, northing values, no?

strump replied:
I did refactoring to use 'int' variable as long as possible. Now `MGRStoLatLon` takes int arguments.

vng commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,251 @@
+            break;
+    }
+
+    return pos;
```
txt.find_first_no_of(" \t\r");

strump replied:
Fixed using `find_first_not_of`

vng commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,251 @@
+    if (eastingStr.size() < 5)
+    {
+        int decShift = 5 - eastingStr.size();
+        easting *= pow(10L, decShift);
```
math::PowUint

strump replied:
Fixed

vng commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,251 @@
+    if (northingStr.size() < 5)
+    {
+        int decShift = 5 - northingStr.size();
+        northing *= pow(10L, decShift);
```
math::PowUint

strump replied:
Fixed

strump commented on android/src/app/organicmaps/widget/placepage/CoordinatesFormat.java:
```diff
@@ -2,23 +2,39 @@
+    return label;
+  }
+
+  public boolean getShowLabel()
```
Renamed method

strump commented on base/string_utils.cpp:
```diff
@@ -408,6 +408,33 @@ std::string to_string_dac(double d, int dac)
+  ss << abs(l);
+
+  return ss.str();
+}
```
It works. Added special case for negative numbers.

biodranik commented on base/string_utils.cpp:
```diff
@@ -408,6 +408,15 @@ std::string to_string_dac(double d, int dac)
+std::string to_string_width(long l, int width)
+{
+  std::ostringstream ss;
+  if (l<0)
```
```suggestion
  if (l < 0)
```

strump replied:
Fixed

biodranik commented on base/string_utils.cpp:
```diff
@@ -408,6 +408,15 @@ std::string to_string_dac(double d, int dac)
   return ss.str();
 }
 
+std::string to_string_width(long l, int width)
```
Should it be a template that accepts all integer types and fails for float types? Or should floating point numbers also be supported?

strump replied:
Refactored function to accept any integer type.

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+
+} UTMPoint;
+
+UTMPoint latlon_to_utm(double lat, double lon);
```
Are these definitions really necessary? It will be harder to edit their signature in two places when necessary.

strump replied:
Refactored functions. Removed forward declarations. Reordered functions.

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+double square_char_to_northing(char n, int set);
+double zone_to_min_northing(char zoneLetter);
+
+const double PI = 3.14159165358979;
```
Please use constexpr everywhere where possible.

biodranik replied:
math::pi ?

strump replied:
Added `constexpr` all over the code. Replaced `PI` with `math::pi`.

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+double zone_to_min_northing(char zoneLetter);
+
+const double PI = 3.14159165358979;
+const double K0 = 0.9996;
```
Please add comments on what it means or a link to the source of this code.

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+  return deg * DEGREES_TO_RADIANS;
+}
+
+double radToDeg(double rad)
```
base::RadToDeg?

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+const int SET_ORIGIN_ROW_LETTERS[] = { 'A', 'F', 'A', 'F', 'A', 'F' };
+
+
+double degToRad(double deg)
```
base::DegToRad?

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+UTMPoint latlon_to_utm(double lat, double lon);
+int latlon_to_zone_number(double lat, double lon);
+char latitude_to_zone_letter(double lat);
+int zone_number_to_central_longitude(int zone_number);
```
These function names do not follow our code convention.
```suggestion
int ZoneNumberToCentralLongitude(int zone_number);
```

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+string FormatUTM(double lat, double lon)
+{
+  if (lat <= -80 || lat > 84)
+    return "Latitude limit exceeded";
```
Why `std::optional<string>` can't be used here and in other similar places?

strump replied:
Added `std::optional<string>` to all functions and JNI calls.

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+  double n = R / sqrt(1 - E * lat_sin * lat_sin);
+  double c = E_P2 * lat_cos * lat_cos;
+
+  double a = lat_cos * mod_angle(lon_rad - central_lon_rad);
```
All const variables should be marked as const.

strump replied:
Fixed.

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+// Generate UTM string from UTM point parameters.
+string utm_to_str(UTMPoint point)
+{
+  return to_string(point.zone_number) + string(1, point.zone_letter) + " " +\
```
```suggestion
  return to_string(point.zone_number) + string(1, point.zone_letter) + " " + \
```

strump replied:
Fixed.

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+string utm_to_mgrs_str(UTMPoint point, int precision)
+{
+  if (point.zone_letter == 'Z')
+    return "Latitude limit exceeded";
```
ditto

strump replied:
Replaced with `return nullopt;`

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+// Convert UTM parameters to lat,lon for WSG 84 ellipsoid.
+// If UTM parameters are valid lat and lon references are used to output calculated coordinates.
+// Otherwise function returns 'false'.
+bool UTMtoLatLon(double easting, double northing, int zone_number, char zone_letter, double &lat, double &lon)
```
```suggestion
bool UTMtoLatLon(double easting, double northing, int zone_number, char zone_letter, double & lat, double & lon)
```

Can std::optional<ms::LatLon> be used for return type?

strump replied:
Fixed formatting

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+
+  double latitude = (p_rad - (p_tan / r) *
+                    (d2 / 2.0 -
+                     d4 / 24.0 * (5.0 + 3.0 * p_tan2 + 10.0 * c - 4.0 * c2 - 9.0 * E_P2)) +
```
Underscored variables do not follow our convention...

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+}
+
+// Convert MGRS parameters to UTM parameters and then use UTM to lat,lon conversion.
+bool MGRStoLatLon(double easting, double northing, int zone_code, char zone_letter, char square_code[2], double &lat, double &lon)
```
ditto for return type?

strump replied:
No more `bool` return type. This function now returns `optional<LatLon>`.

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+    if (curCol > 'Z')
+    {
+      if (rewindMarker)
+        return -1;
```
`return std::optional<double>`?

biodranik commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,567 @@
+double square_char_to_northing(char n, int set) {
+
+  if (n > 'V')
+    return -1;
```
ditto

biodranik replied:
Or if you find std::optional too heavy, please introduce kInvalid... constant instead.

biodranik commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,216 @@
+size_t MatchZoneCode(string const & query, int &zone_code)
+{
+  auto pos = query.find_first_not_of(kSpaceChars);
+  if (query.size()-pos < 2)
```
```suggestion
  if (query.size() - pos < 2)
```

biodranik commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,216 @@
+// Matches 2 digits zone code. Returns end position of matched chars or string::npos if no match.
+size_t MatchZoneCode(string const & query, int &zone_code)
+{
+  auto pos = query.find_first_not_of(kSpaceChars);
```
const?

strump replied:
Added `const` all over the code.

biodranik commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,216 @@
+}
+
+// Matches 2 digits zone code. Returns end position of matched chars or string::npos if no match.
+size_t MatchZoneCode(string const & query, int &zone_code)
```
```suggestion
size_t MatchZoneCode(string const & query, int & zone_code)
```
There are many similar typos.

strump replied:
Fixed formatting.

biodranik commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,216 @@
+    return string::npos;
+
+  char dig1 = query[pos];
+  char dig2 = query[pos+1];
```
const?
pos + 1 ?

biodranik commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,216 @@
+
+
+// Matches zone letter ignoring spaces. Returns end position of matched chars or string::npos if no match.
+size_t MatchZoneLetter(string const & query, char &zone_letter, size_t startPos)
```
```suggestion
size_t MatchZoneLetter(string const & query, char & zone_letter, size_t startPos)
```

strump replied:
Fixed formatting.

biodranik commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,216 @@
+  if (query.size() == pos)
+    return string::npos;
+
+  char l = query[pos];
```
const...

strump replied:
Fixed.

biodranik commented on search/utm_mgrs_coords_match.cpp:
```diff
@@ -0,0 +1,216 @@
+    char ch = query[pos];
+    if (ch >= '0' && ch <= '9')
+    {
+      n = n*10 + (ch-'0');
```
```suggestion
      n = n * 10 + (ch - '0');
```
That's easier to read.

strump replied:
Fixed formatting.

vng commented:
@strump If you took this code from any open source, I'd prefer leave formatting as-is in cpp module.

- Yes, better to make ```optional<LatLon> Foo() instead of bool Foo(double & lat, double & lon```).
- Me don't like ```optional<double>``` :) if we can make -1 invalid constant.
- can keep empty string in case of SmthToString error?

biodranik commented:
If this code was copied from open source, then 1) it should have a proper attribution here and 2) we should update data/copyright.html
(of course, if the license was not copyleft).

strump commented:
Did some refactoring. Introduced `std::optional<std::string>` and `std::optional<ms::LatLon>` types. Added `const` and `constexpr` all over the code.

Should I squash commits and split into multiple commits: Platform, Android, Search, xCode?

vng commented on search/processor.cpp:
```diff
@@ -608,15 +609,28 @@ bool Processor::SearchCoordinates()
+  {
+    coords_found = true;
+    auto latLon = maybeLatLon.value();
+    results.emplace_back(latLon.m_lat, latLon.m_lon);
```
```
auto ll = ...;
if (ll)
  emplace_back(ll->m_lat, ll->m_lon);
```

strump replied:
Fixed. Using `->` operator could be little slower 'cause it unpacks optional. But this code is not used often.

vng commented on platform/platform_tests/utm_mgrs_utils_tests.cpp:
```diff
@@ -0,0 +1,70 @@
+
+using namespace utm_mgrs_utils;
+
+#define OPT(a) std::make_optional(a)
```
Well, better to make shortcut function ```std::optional<std::string> OPT(std::string s) { return s; }```

But, I'd prefer to avoid ```optional<string>``` if we have an empty string, no?

vng commented on platform/utm_mgrs_utils.cpp:
```diff
@@ -0,0 +1,525 @@
+using namespace base; // To use base::DegToRad and base::RadToDeg
+using namespace ms; // To use ms::LatLon
+
+typedef struct UTMPoint_Value
```
just ```struct UTMPoint```

strump replied:
Fixed struct definition.

vng commented on search/search_tests/utm_mgrs_coords_match_test.cpp:
```diff
@@ -0,0 +1,156 @@
+{
+  TEST(maybeLatLon.has_value(), ());
+
+  auto const actualLat = maybeLatLon.value().m_lat;
```
maybeLatLon->m_lat

vng commented on search/search_tests/utm_mgrs_coords_match_test.cpp:
```diff
@@ -0,0 +1,156 @@
+  auto const actualLat = maybeLatLon.value().m_lat;
+  TEST(base::AlmostEqualAbsOrRel(actualLat, expectedLat, kEps), ("Lat is not close", actualLat, expectedLat));
+
+  auto const actualLon = maybeLatLon.value().m_lon;
```
maybeLatLon->m_lon

vng reviewed (commented):
Well, LGTM here, but

I'd prefer to use std::string and an empty string if error, instead of ```optional<string>```

biodranik commented:
`optional<string>` should be used when empty string is still a valid result. If it's invalid, then returning empty `string` is ok.

telluroid commented:
Hi @strump,
I just saw your work on this and would like to mention two things that came up my mind.

You said that:
> Some coordinates are not convertible to UTM or MGRS

This is not true. Both UTM and MGRS can be calculated for the whole globe. See for example: https://en.wikipedia.org/wiki/Military_Grid_Reference_System

In your Screenshots the last two digits of UTM and MGRS are different. This is strange because the last 5 digits of the easting and northing should be exactly the same.

vng reviewed (approved):
LGTM. Squash! all commits into one when merge.

biodranik commented:
> when OrganicMaps will show 3D globe I'll upgrade UTM/MGRS formulas to support all latitudes.

Is it mentioned somewhere in the code?

biodranik reviewed (approved):
Thank you! Please update or create an issue to finish coordinates support for iOS UI.
